### ISSUE TITLE
The application crashes while using DrawPDFPage for specific document in DidEnterBackground method.

### ISSUE THREAD

**Issue by manojSF3414** [xamarin/xamarin-macios#21167]
### Apple platform

iOS

### Framework version

Other

### Affected platform version

VS for Mac 2019 v8.10.25, Xamarin.iOS v16.4.0.23 

### Description

I’m currently using CGPDFPage and CGContext to convert a PDF into an image. The process works fine in most cases, but I’ve encountered an issue when the PDF file is larger than 50 MB and the application is sent to the background by pressing the home button and then reopened. This problem only occurs on physical devices; emulators work without any issues.

We are using the following code for this conversion:

```
fileStream.CopyTo(memoryStream);
CGDataProvider provider = new CGDataProvider(memoryStream.ToArray());
CGPDFDocument cGPDFDocument = null;
 image = null;
    cGPDFDocument = new CGPDFDocument(provider);
using(CGPDFPage pdfPage = cGPDFDocument?.GetPage(1))
{
     if (pdfPage != null)
    {
        CGRect rect = pdfPage.GetBoxRect(CGPDFBox.Media);
        if (pdfPage.RotationAngle == 90 || pdfPage.RotationAngle == 270)
        {
            rect = new CGRect(rect.X, rect.Y, rect.Height, rect.Width);
        }
        nfloat factor = (nfloat)0.5;
        CGRect bounds = new CGRect(rect.X * factor, rect.Y * factor, rect.Width * factor, rect.Height * factor);
        UIGraphics.BeginImageContext(bounds.Size);
        CGContext context = UIGraphics.GetCurrentContext();
        context.SetFillColor(1.0f, 1.0f, 1.0f, 1.0f);
        context.FillRect(bounds);
        context.TranslateCTM(0, bounds.Height);
        context.ScaleCTM(factor, -factor);
        context.ConcatCTM(pdfPage.GetDrawingTransform(CGPDFBox.Crop, rect, 0, true));
        context.SetRenderingIntent(CGColorRenderingIntent.Default);
        context.InterpolationQuality = CGInterpolationQuality.Default;
        context.DrawPDFPage(pdfPage);
        image = UIGraphics.GetImageFromCurrentImageContext();
        UIGraphics.EndImageContext();
        imageView.Image = image; 
        parentView.AddSubview(imageView);
    }
}
```

We’ve created a simple Xamarin.iOS application to replicate this issue, with the problematic PDF included in the Assets. You can download the sample here: https://www.syncfusion.com/downloads/support/directtrac/general/ze/PdfToImageSample-556012354

We’ve also recorded the issue for reference, which you can view here: https://www.syncfusion.com/downloads/support/directtrac/general/ze/CrashBGImgRec-1376481150

The issue consistently occurs in the line `context.DrawPDFPage(pdfPage);` in my application. In the provided sample, the issue happens consistently when a breakpoint is placed on that line, and it can also occur randomly without a breakpoint.

This function is being invoked in the DidEnterBackground override method of AppDelegate.

Do we need to set any specific configurations to properly retrieve the image from CGPDFPage and add it to the context?

### Steps to Reproduce

1.  Open the application at the following link: https://www.syncfusion.com/downloads/support/directtrac/general/ze/PdfToImageSample-556012354
2. Build and run the application on the physical device.
3. make the application to enter the background by pressing the home button 
4. Open the application again.
5. Application crashes at "context.DrawPDFPage(pdfPage);"

### Did you find any workaround?

No, not yet

**Comment by rolfbjarne** [comment]
The pdf renders on my iPad Pro 3, which device are you trying on? It might be a memory issue, where the app ends up consuming too much memory and is killed.

Can you see if there are any crash reports (https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#crash-reports) or if anything is printed to the device log (https://support.apple.com/en-in/guide/console/cnsl1012/mac)?

**Comment by manojSF3414** [comment]
Hi @rolfbjarne 
I tested the code on the iPad (5th gen) with iPadOS 16.7.8 and iPad Pro (11 inches) with iPadOS 17.6.1 and faced the same issue. Could you please try debugging the code where you can consistently replicate the issue? This issue is consistently replicated in our project and rarely in the provided solution. 
Log message:
![image](https://github.com/user-attachments/assets/e7e34347-5f7f-452a-92f8-0f32b66ce745)
Device Logs
[GettingStartediOS-ErrorLog.zip](https://github.com/user-attachments/files/16963741/GettingStartediOS-ErrorLog.zip)

**Comment by rolfbjarne** [comment]
> [GettingStartediOS-ErrorLog.zip](https://github.com/user-attachments/files/16963741/GettingStartediOS-ErrorLog.zip)

That's because you triggered a watchdog, something took too long:

> exhausted real (wall clock) time allowance of 10.00 seconds

From the stack trace it looks like the main thread is waiting on a lock, but unfortunately the important symbols have been stripped away, so there's just memory addresses.

This can be improved by adding the following to your csproj, which will disable native symbol stripping, and thus provide better stack traces in crash reports:

```xml
<PropertyGroup>
    <NoSymbolStrip>true</NoSymbolStrip>
</PropertyGroup>
```

**Comment by rolfbjarne** [comment]
> I tested the code on the iPad (5th gen) with iPadOS 16.7.8 and iPad Pro (11 inches) with iPadOS 17.6.1 and faced the same issue. Could you please try debugging the code where you can consistently replicate the issue? This issue is consistently replicated in our project and rarely in the provided solution.

Your project is a Xamarin.iOS project, and support for Xamarin.iOS ended [earlier this year](https://dotnet.microsoft.com/en-us/platform/support/policy/xamarin).

I previously tested your project real quick to see if anything was obviously wrong somewhere, but I didn't find anything. A more involved support is out of scope for a product that's been discontinued already.

If you upgrade your project to .NET, we'll be more than happy to investigate deeper any issues you run into (the problem you're running into might not even happen anymore).

**Comment by manojSF3414** [comment]
Hi @rolfbjarne 
I have added <NoSymbolStrip>true</NoSymbolStrip> and got the following details
[GettingStartediOS-2024-09-18-201323.zip](https://github.com/user-attachments/files/17047942/GettingStartediOS-2024-09-18-201323.zip)
![image](https://github.com/user-attachments/assets/729459e6-960b-4e34-97c2-e8e0d9f97ac1)
![image](https://github.com/user-attachments/assets/ebef01f0-4863-44c2-963d-c1a07305b9cf)
![image](https://github.com/user-attachments/assets/347755e3-ca3f-4709-9a16-d3e482454d92)
![image](https://github.com/user-attachments/assets/3928a407-f65f-4601-aaed-1f96e12df877)

If this is a memory issue that happens on physical devices only how can we handle this memory pile up issue only with certain PDF documents?

**Comment by rolfbjarne** [comment]
From the crash report:

> exhausted real (wall clock) time allowance of 10.00 seconds

iOS terminated your app because it was stuck doing something for too long.

From the stack trace it looks like it took too long to render the pdf.

Maybe it would be possible to render it to an image on a background thread, and then just render the image on the main thread?
